src/terminal/input/terminalInput.cpp

```diff
@@ -515,169 +162,442 @@ TerminalInput::OutputType TerminalInput::HandleKey(const INPUT_RECORD& event)
+        if (altGrIsPressed && (bothCtrlsArePressed || (rightAltIsPressed && leftCtrlIsReallyPressed)))
         {
-            // Pressing the control key causes all bits but the 5 least
-            // significant ones to be zeroed out (when using ASCII).
-            return _makeEscapedOutput(ctrlAltChar & 0b11111);
+            unicodeChar = _makeCtrlChar(unicodeChar);
         }
-
-        // Currently, when we're called with Alt+Ctrl+@, ch will be 0, since Ctrl+@ equals a null byte.
-        // VkKeyScanW(0) in turn returns the vkey for the null character (ASCII @).
-        // -> Use the vkey to determine if Ctrl+@ is being pressed and produce ^[^@.
-        if (ch == UNICODE_NULL && vkey == LOBYTE(OneCoreSafeVkKeyScanW(0)))
+        auto charSequence = _makeCharOutput(unicodeChar);
+        if (!charSequence.empty())
         {
-            return _makeEscapedOutput(L'\0');
+            // We may also need to apply an Alt prefix to the char sequence, but
+            // if this is an AltGr key, we only do so if both Alts are pressed.
+            const auto bothAltsArePressed = WI_AreAllFlagsSet(controlKeyState, ALT_PRESSED);
+            if (altGrIsPressed ? bothAltsArePressed : altIsPressed)
+            {
+                charSequence = _makeEscapedOutput(charSequence);
+            }
         }
+        return charSequence;
+    }
+
+    // If we don't have a UnicodeChar, we'll try and determine what the key
+    // would have transmitted without any Ctrl or Alt modifiers applied. But
+    // this only makes sense if there were actually modifiers pressed.
+    if (!altIsPressed && !ctrlIsPressed)
+    {
+        return MakeOutput({});
     }
 
-    // If a modifier key was pressed, then we need to try and send the modified sequence.
-    if (WI_IsAnyFlagSet(keyEvent.dwControlKeyState, MOD_PRESSED))
+    // We need the current keyboard layout and state to lookup the character
+    // that would be transmitted in that state (via the ToUnicodeEx API).
+    const auto hkl = GetKeyboardLayout(GetWindowThreadProcessId(GetForegroundWindow(), nullptr));
+    auto keyState = _getKeyboardState(virtualKeyCode, controlKeyState);
+    const auto flags = 4u; // Don't modify the state in the ToUnicodeEx call.
+    const auto bufferSize = 16;
+    auto buffer = std::array<wchar_t, bufferSize>{};
+
+    // However, we first need to query the key with the original state, to check
+    // whether it's a dead key. If that is the case, ToUnicodeEx should return a
+    // negative number, although in practice it's more likely to return a string
+    // of length two, with two identical characters. This is because the system
+    // sees this as a second press of the dead key, which would typically result
+    // in the combining character representation being transmit twice.
+    auto length = ToUnicodeEx(virtualKeyCode, 0, keyState.data(), buffer.data(), bufferSize, flags, hkl);
+    if (length < 0 || (length == 2 && buffer.at(0) == buffer.at(1)))

```

Comment:
`length` may also be 4 if it's a surrogate pair, no? How should we handle that?

Reply:
I guess it's theoretically possible, but none of the keyboard layouts on the [kbdlayout](https://www.kbdlayout.info/) site appear to have surrogate-pair dead keys. If there is such a thing, it's got to be rare, so I'd be inclined to ignore the problem until someone complains about it.